src/pools.jl

```diff
@@ -0,0 +1,109 @@
+module Pools
+
+import ..ConcurrentStack, ..Lockable
+
+export Pool, acquire, release
+
+import Base: Semaphore, acquire, release
+
+"""
+    Pool(T; max::Int=typemax(Int))
+    Pool(K, T; max::Int=typemax(Int))
+
+A threadsafe object for managing a pool of objects of type `T`, optionally keyed by objects
+of type `K`. Objects can be requested by calling `acquire(f, pool, [key])`, where `f` is a
+function that returns a new object of type `T`.
+The `key` argument is optional and can be used to lookup objects that match a certain criteria.
+If no pool exists for the given key, one will be created.
+
+If `max` is specified, the pool will limit the number of objects
+that can be checked out at any given time. If the limit has been reached, `acquire` will
+block until an object is returned to the pool via `release`.
+
+By default, `release(pool, obj)` will return the object to the pool for reuse.
+`release(pool)` will return the "permit" to the pool while not returning
+any object for reuse.
+"""
+struct Pool{K, T}
+    sem::Semaphore
+    values::Lockable{Dict{K, ConcurrentStack{T}}}
+end
+
+Pool(T; max::Int=typemax(Int)) = Pool{Nothing, T}(Semaphore(max), Lockable(Dict{Nothing, ConcurrentStack{T}}()))
+Pool(K, T; max::Int=typemax(Int)) = Pool{K, T}(Semaphore(max), Lockable(Dict{K, ConcurrentStack{T}}()))
+
+Base.empty!(pool::Pool) = Base.@lock pool.values empty!(pool.values[])
+
+TRUE(x) = true
+
+"""
+    acquire(f, pool::Pool{K, T}, [key::K]; forcenew::Bool=false, isvalid::Function) -> T
+
+Get an object from a `pool`, optionally keyed by the provided `key`. If no pool exists for the given key, one will be created.
+The provided function `f` must create a new object instance of type `T`.
+The acquired object MUST be returned to the pool by calling `release(pool, key, obj)` exactly once.
```

Comment:
Do we need to call 3-arg `release`? isn't it just:
```suggestion
Each `acquire` call MUST be matched by exactly one `release` call.
```